## c++ 32位异常还原

首先确定try的位置

![image-20230904000059407](imgs\image-20230904000059407.png)

上面明显是一个SEH结构，在c++异常中，state固定在var_4的位置上，这里state初始化位-1，我们将var_4改名为state

![image-20230904000432664](imgs\image-20230904000432664.png)

上图为`ida`的反编译图,当state赋值为0时，为try的开始，state赋值为-1时表示try块的结束

下面我们分析每一个抛出的异常

![image-20230904000758938](imgs\image-20230904000758938.png)

_CxxThrowException的第一个参数为抛出异常的地址，那么这里的异常应该时一个常量3，我们可以跳转到第二个参数ThrowInfo

![image-20230904001121066](imgs\image-20230904001121066.png)

![7b052f9f45bb8742b0bbec672552143](imgs\7b052f9f45bb8742b0bbec672552143.jpg)

上面是ThrowInfo的数据和结构

可以看到这个异常时一个简单类型异常，并且为int

case 1 - 4 都是差不多的，它们可还原为下面的代码

~~~c++
		case 1:
    		throw 3;
   		case 2:
     		throw 3.0f;
    	case 3:
     		throw '3';
    	case 4:
     		throw 3.0;
~~~

